File: frontend/src/model/ProblemDetails.ts
export type ProblemDetails =  {
    type: string, // "https://example.com/probs/out-of-credit",
    title: string, //"You do not have enough credit.",
    detail?: string | null, //"Your current balance is 30, but that costs 50.",
    instance: string, // "/account/12345/msgs/abc",
}

File: src/main/kotlin/pl/ejdev/model/ProblemDetails.kt
package pl.ejdev.model

import io.ktor.server.plugins.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class ProblemDetails(
    val type: String, // "https://example.com/probs/out-of-credit",
    val title: String, //"You do not have enough credit.",
    val detail: String?, //"Your current balance is 30, but that costs 50.",
    val instance: String, // "/account/12345/msgs/abc",
) {
    companion object {
        fun from(
            call: RoutingCall,
            title: String,
            detail: String?,
        ): ProblemDetails = ProblemDetails(
            type = "${call.request.origin.scheme}://${call.request.headers["host"]}${call.request.origin.uri}",
            title = title,
            detail = detail,
            instance = call.request.origin.uri
        )
    }
}
